clients/backup/backup.go

```diff
@@ -0,0 +1,283 @@
+package backup
+
+import (
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"mime"
+	"mime/multipart"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/influxdata/influx-cli/v2/api"
+	"github.com/influxdata/influx-cli/v2/clients"
+)
+
+type Client struct {
+	clients.CLI
+	api.BackupApi
+
+	// Local state tracked across steps in the backup process.
+	baseName       string
+	bucketMetadata []api.BucketMetadataManifest
+	manifest       Manifest
+}
+
+type Params struct {
+	// Organization containing TSM data to back up.
+	// If not set, all orgs will be included.
+	OrgID string
+	Org   string
+
+	// Bucket containing TSM data to back up.
+	// If not set, all buckets within the org filter will be included.
+	BucketID string
+	Bucket   string
+
+	// Path to the directory where backup files should be written.
+	Path string
+
+	// Compression to use for local copies of snapshot files.
+	Compression FileCompression
+}
+
+func (p *Params) matches(bkt api.BucketMetadataManifest) bool {
+	if p.OrgID != "" && bkt.OrganizationID != p.OrgID {
+		return false
+	}
+	if p.Org != "" && bkt.OrganizationName != p.Org {
+		return false
+	}
+	if p.BucketID != "" && bkt.BucketID != p.BucketID {
+		return false
+	}
+	if p.Bucket != "" && bkt.BucketName != p.Bucket {
+		return false
+	}
+	return true
+}
+
+const backupFilenamePattern = "20060102T150405Z"
+
+func (c *Client) Backup(ctx context.Context, params *Params) error {
+	if err := os.MkdirAll(params.Path, 0777); err != nil {
+		return err
+	}
+	c.baseName = time.Now().UTC().Format(backupFilenamePattern)
+
+	if err := c.downloadMetadata(ctx, params); err != nil {
+		return fmt.Errorf("failed to backup metadata: %w", err)
+	}
+	if err := c.downloadBucketData(ctx, params); err != nil {
+		return fmt.Errorf("failed to backup bucket data: %w", err)
+	}
+	if err := c.writeManifest(params); err != nil {
+		return fmt.Errorf("failed to write backup manifest: %w", err)
+	}
+	return nil
+}
+
+// downloadMetadata downloads a snapshot of the KV store, SQL DB, and bucket
+// manifests from the server. KV and SQL are written to local files. Bucket manifests
+// are parsed into a slice for additional processing.
+func (c *Client) downloadMetadata(ctx context.Context, params *Params) error {
+	log.Println("INFO: Downloading metadata snapshot")
+	rawResp, err := c.GetBackupMetadata(ctx).AcceptEncoding("gzip").Execute()
+	if err != nil {
+		return fmt.Errorf("failed to download metadata snapshot: %w", err)
+	}
+
+	kvName := fmt.Sprintf("%s.bolt", c.baseName)
```

Comment:
We are setting a base name on the server as well and included it with the response (https://github.com/influxdata/influxdb/blob/40c7d694154e75103f614cc15fdf329cdb5c0bcc/http/backup_service.go#L163). Looking at the client implementation, I think it's better to set it here though - easier to make sure all the backed up files match.

Unless you need it for something on the client (I don't think so?), at some point we might want to remove setting the filename on the server response, just to keep things from being confusing.

Reply:
:+1: I think removing it would make sense. I opened an issue to support gzipping the response on the server-side, I'll add a note about dropping the file names there as well.